### PR TITLE
Enable or disable the plane transport

### PR DESCRIPTION
**Pull request by magopian**
[user story](https://www.notion.so/TRANSPORTS-Ajouter-du-transport-en-avion-par-exemple-en-ajoutant-la-mangue-5d57dee47db04f42b26776a1fbc77922)
[user story (bis)](https://www.notion.so/fd0068a7db8348c584cd53a3eea399f2?v=90c1a9d44ff845c090b6f491065ba2ef&p=2786364aac184ee88a14b480045ebe61&pm=s)

Following #218

TODO:
- [x] add an argument in the API to specify whether or not the transport is by plane
- [x] only allow plane selection for type4 ingredients (coming from out of Maghred and Europe by plane)
- [ ] modifications UI/UX avec Laurent
- [x] fix tests
- [x] fix failing (commented out) test in [server.spec.js](https://github.com/MTES-MCT/ecobalyse/pull/220/files#diff-1177cee604b4d39be39ced46f1b1e6e2c9af668d863d4072994ce6dc07080807R333-R340)

### CODE REVIEW

**Comment by n1k0** [comment]
There has been discussions about the UX proposed by this patch at the standup today, we agreed Laurent should have a look at how the options could be better presented. Also, Lisa suggested plane checkboxes should only be exposed for category 4 ingredients (out of EU and Maghreb by plane).

File: src/Data/Food/Builder/Query.elm [hunk 1]
```diff
@@ -92,25 +93,29 @@ carrotCake =
           , mass = Mass.grams 120
           , variant = Default
           , country = Nothing
+            , byPlane = False
```
Comment (n1k0) [inline src/Data/Food/Builder/Query.elm:96]: Is it expected indentation isn't consistant here?

File: src/Data/Food/Builder/Recipe.elm [hunk 2]
```diff
@@ -213,13 +214,12 @@ computeIngredientTransport db { ingredient, country, mass } =
                     db.transports
                         |> Transport.getTransportBetween Scope.Food emptyImpacts code france
                         |> (\ingredientTransport ->
-                                if ingredient.defaultOrigin == Origin.OutOfEuropeAndMaghrebByPlane then
-                                    -- Special case: if the default origin of an ingredient is "by plane",
-                                    -- then we take an air transport ratio of 1
+                                if (ingredient.defaultOrigin == Origin.OutOfEuropeAndMaghrebByPlane) && byPlane then
+                                    -- Special case: if the default origin of an ingredient is "by plane"
+                                    -- and we selected a transport by plan, then we take an air transport ratio of 1
```
Comment (n1k0) [inline src/Data/Food/Builder/Recipe.elm:219]: s/by plan/by plane

File: src/Data/Food/Ingredient.elm [hunk 3]
```diff
@@ -37,6 +38,11 @@ type Id
     = Id String
 
 
+byPlaneFromOrigin : Origin.Origin -> Bool
+byPlaneFromOrigin origin =
+    origin == Origin.OutOfEuropeAndMaghrebByPlane
```
Comment (n1k0) [inline src/Data/Food/Ingredient.elm:43]: Nit: I'd rather have the full ingredient passed as the argument, and maybe the function renamed: 

```elm
if Ingredient.byPlaneByDefault mango then
``` 

reads a little better than 

```elm
if Ingredient.byPlaneFromOrigin mango.defaultOrigin then
```

File: src/Data/Food/Builder/Query.elm [hunk 4]
```diff
@@ -92,25 +93,29 @@ carrotCake =
...
           }
         , { id = Ingredient.idFromString "wheat"
           , name = "Blé tendre"
           , mass = Mass.grams 140
           , variant = Default
           , country = Nothing
+            , byPlane = False
```
Comment (n1k0) [inline src/Data/Food/Builder/Query.elm:103]: The same way we have `Maybe Country` helping identifying when no option is explicitely passed in an query (especially with the API), we may probably want a `Maybe Bool` here. Also that would keep things consistent.

I understand it's complicated. Basically, `Maybe a` will help you understand better if the user has explicitly overridden a default or not; `Nothing` means the query will rely on default value, whatever it is: it doesn't know about it. `Just "x"` means the user have explicitely set the value to `"x"`. So when we serialize the query, we can distinguish if the user has overridden a default or not. If we enforce a value being always present, you can't easily know if the user has overridden it or not. And if we ever change the default values, thery're not hardcoded in serialized/saved queries (eg. bookmarks or api call urls).

I bet it's even more confusing now. Available to chat about it.

File: src/Data/Food/Builder/Query.elm (hunk 1 above)
Comment (magopian) [reply src/Data/Food/Builder/Query.elm:96]: no idea how that happened, fixing this 👍

File: src/Data/Food/Builder/Query.elm (hunk 4 above)
Comment (magopian) [reply src/Data/Food/Builder/Query.elm]: mmmm, I'm having some "impossible states" vibes: in theory i could have a mango (`defaultOrigin` is `OutOfEuropeAndMaghrebByPlane`) that has `country = Nothing` (so should use its default origin), and `byPlane = False` (or even `byPlane = Just False`).

I guess this means there's an issue with data modeling... any ideas?

Comment (magopian) [reply src/Data/Food/Builder/Query.elm]: I'll go with it for now, we'll see if this needs a refactor in the future. For now the validation of correct data is done in `Recipe.ingredientFromQuery` and I added a test to make sure this is done.

**Comment by magopian** [comment]
Ok I added an tentative API (I know it's ugly, we might want to refactor the API to stop doing query string parsing and use json payloads instead as discussed with @n1k0). Mind having a look again?

There's a commented out test that doesn't fail as it should, haven't had time to investigate yet (it returns a different error format for some reason).

File: src/Page/Food/Builder.elm [hunk 5]
```diff
@@ -581,7 +583,29 @@ updateIngredientFormView { excluded, db, ingredient, impact, transportImpact } =
...
+                        , attribute "role" "switch"
+                        , checked <| ingredientQuery.byPlane == Just True
+                        , disabled <| ingredient.country == Nothing
+                        , onCheck
+                            (\checked ->
+                                event { ingredientQuery | byPlane = Just checked }
+                            )
+                        ]
+                        []
+                    , text " par avion)"
```
Comment (n1k0) [inline src/Page/Food/Builder.elm:603]: Note: Still quite unsure about this UI, we could rather have a toggler using a plane icon or something. That may be addressed in a next PR though, where Laurent would be involved.

**Review by n1k0 — COMMENTED** [review]
A bunch of supplementary comments

File: src/Server/Query.elm [hunk 6]
```diff
@@ -171,6 +192,19 @@ packagingParser packagings string =
             Err <| "Format d'emballage invalide : " ++ string ++ "."
 
 
+validateBool : String -> Result String Bool
+validateBool str =
+    case str of
+        "true" ->
+            Ok True
```
[the remark this answers is not in the document]
Comment (magopian) [reply src/Server/Query.elm:207]: we can use `byPlane` and `default`, sure. Not sure what `default` would bring for now, as it'll be the same as not defining it (and we don't have extra parameters... yet ;)

File: src/Page/Food/Builder.elm (hunk 5 above)
Comment (magopian) [reply src/Page/Food/Builder.elm:603]: Sure, it's part of the "todo list" in this PR 👍

File: src/Server/Query.elm (hunk 6 above)
[the remark this answers is not in the document]
Comment (magopian) [reply src/Server/Query.elm:207]: current status : 😬 
```
        - `mango;123;default` signifie *123g de mangues en provenance de son origine par défaut (hors Europe-Maghreb, transport par avion)*;
        - `mango;123;default;BR` signifie *123g de mangues en provenance du Brésil, par avion*;
        - `mango;123;default;BR;byPlane` signifie de la même manière *123g de mangues en provenance du Brésil, par avion*;
        - `mango;123;default;BR;default` signifie toujours *123g de mangues en provenance du Brésil, par avion*;
        - `mango;123;default;BR;notByPlane` signifie *123g de mangues en provenance du Brésil, par bateau (pas par avion)*;
```

[the remark this answers is not in the document]
Comment (n1k0) [reply src/Server/Query.elm:207]: Quand le métier est compliqué, la doc est compliquée 🤷

**Comment by magopian** [comment]
Ok, a bunch of new changes, what do you think @n1k0 ? (this has been a difficult one 😬 )

File: openapi.yaml [hunk 7]
```diff
@@ -629,7 +629,9 @@ components:
         - `tomato;123;default;ES` signifie *123g de tomates bio en provenance d'Espagne*;
         - `mango;123;default` signifie *123g de mangues en provenance de son origine par défaut (hors Europe-Maghreb, transport par avion)*;
         - `mango;123;default;BR` signifie *123g de mangues en provenance du Brésil, par avion*;
-        - `mango;123;default;BR;false` signifie *123g de mangues en provenance du Brésil, par bateau (pas par avion)*;
+        - `mango;123;default;BR;byPlane` signifie de la même manière *123g de mangues en provenance du Brésil, par avion*;
+        - `mango;123;default;BR;default` signifie toujours *123g de mangues en provenance du Brésil, par avion*;
+        - `mango;123;default;BR;notByPlane` signifie *123g de mangues en provenance du Brésil, par bateau (pas par avion)*;
```
Comment (n1k0) [inline openapi.yaml:634]: Toujours pas entièrement à l'aise avec cette liste, j'ai essayé de les reformuler de façon plus claire :

- `mango;123;default;BR`, `mango;123;default;BR;byPlane` et `mango;123;default;BR;default` sont équivalents et signifient *123g de mangues en provenance du Brésil, acheminés par avion* (mode de transport par défaut)
- `mango;123;default;BR;notByPlane` permet de désactiver le mode de transport par avion et utiliser le mode d'acheminement classique, par défaut le bateau pour le Brésil;

Mais je ne suis toujours pas convaincu, c'est trop complexe, comme les règles métiers correspondantes d'ailleurs. Il faudrait aussi préciser que seuls les ingrédients dont le transport par défaut incluent l'avion peuvent voir ce mode de transport désactivé, sinon une erreur sera renvoyée.

J'aurais tendance à suggérer du pair avec Pascal et/ou Lisa pour qu'iels observent concrètement la complexité de l'interface (*interface* au sens *jeu de paramètres décrivant cette partie de la requête*)  et de sa documentation pour éventuellement simplifier ou se raviser.

**Review by n1k0 — APPROVED** [review]
I'm gonna approve the patch as the code is fine but I'm kinda against the added business complexity here, though I understand things need to ship at some point.

File: openapi.yaml (hunk 7 above)
Comment (n1k0) [reply openapi.yaml:634]: D'ailleurs je me rend compte que `notByPlane` pourrait être avantageusement renommé `noPlane`.

Comment (magopian) [reply openapi.yaml:634]: Oui, c'est mieux, je modifie ça ;)

Comment (magopian) [reply openapi.yaml]: Concernant les règles métiers, j'ai essayé de reformuler et de rajouter un lien vers la doc, en espérant que ce soit plus clair ?

Comment (n1k0) [reply openapi.yaml]: C'est mieux mais toujours compliqué, je ne suis pas entièrement certain qu'il soit possible d'obtenir mieux en restant sur les règles métier en question de toute façon.

**Comment by magopian** [comment]
I'm landing this now, let's circle back to it if/when we decide to simplify (or complexify 😱 ) the rules

**Pull request merged**